File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.go
func diagonalSort(mat [][]int) [][]int {
    diagonals := make(map[int][]int)
    
    m, n := len(mat), len(mat[0])
    
    for i, row := range mat {
        for j, val := range row {
            diagonals[j - i] = append(diagonals[j - i], val)
        }
    }
    
    for _, vals := range diagonals {
        sort.Ints(vals)
    }
    
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            mat[i][j] = diagonals[j - i][0]
            diagonals[j - i] = diagonals[j - i][1:]
        }
    }

    return mat
}

File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.py
class Solution:
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        # Build lists of diagonals
        diagonals = defaultdict(list)
        for i, row in enumerate(mat):
            for j, val in enumerate(row):
                diagonals[j - i].append(val)
        
        # Sort the diagonals
        for vals in diagonals.values():
            vals.sort(reverse=True)
        
        # re-insert diagonals into the matrix
        for i, row in enumerate(mat):
            for j, _ in enumerate(row):
                mat[i][j] = diagonals[j - i].pop()
        
        return mat